Site Editor: Create auto-drafts for modified files only

david-szabo97 opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
Only run synchronization logic for modified files rather than all file-based template and template part files.

Fixes https://github.com/WordPress/gutenberg/issues/25868

## How has this been tested?

- Edit a theme block template file
- Load site editor
- Confirm `wp_template` is created or updated
- You might need to reload the site editor once

## Types of changes
Code quality & Performance

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

Addison-Stavlo commented on lib/templates.php:
```diff
@@ -231,3 +224,59 @@ function filter_rest_wp_template_query( $args, $request ) {
+
+	gutenberg_create_template_auto_drafts();
+}
+register_shutdown_function( 'gutenberg_create_template_auto_drafts_on_shutdown' );
```
Curious what the purpose of adding this on shutdown is?

david-szabo97 replied:
The idea is to run it once the response is sent to the user. So we don't delay TTFB.

When we open the site editor we re-create the auto-drafts in the background. If any of the template files changed then they will be updated in the database as well. (See `Reconciliation after updating files`)

Addison-Stavlo reviewed (commented):
Its nice to see more eyes in this template (and part) auto-drafting arena.  It feels a bit all over the place atm (not this PR, just whats in place in general) and seems like it needs to be rethought a bit.

I see its still `In Progress` but a few Qs and thoughts:

david-szabo97 commented on lib/templates.php:
```diff
@@ -202,13 +202,6 @@ function filter_rest_wp_template_collection_params( $query_params ) {
  * @return array Filtered $args.
  */
 function filter_rest_wp_template_query( $args, $request ) {
-	// Create auto-drafts for each theme template files.
-	$block_template_files = gutenberg_get_template_paths();
-	foreach ( $block_template_files as $path ) {
-		$template_type = basename( $path, '.html' );
-		gutenberg_find_template_post_and_parts( $template_type, array( $template_type ) );
-	}
```
It was a quick hack to get it working.

david-szabo97 commented on lib/templates.php:
```diff
@@ -231,3 +224,59 @@ function filter_rest_wp_template_query( $args, $request ) {
+	$last_auto_drafts_versions = get_option( 'gutenberg_last_template_auto_drafts_theme_versions', array() );
+	if ( isset( $last_auto_drafts_versions[ $stylesheet ] ) ) {
+		$last_version = $last_auto_drafts_versions[ $stylesheet ];
+		if ( version_compare( $theme->Version, $last_version, '>' ) ) {
```
This saves the version in a map `key = $stylesheet | value = version` so we can switch between themes without any problems. Or I misunderstood your question?

Addison-Stavlo replied:
Oh no, that makes sense 👍

david-szabo97 commented on lib/templates.php:
```diff
@@ -231,3 +224,59 @@ function filter_rest_wp_template_query( $args, $request ) {
+	}
+	update_option( 'gutenberg_last_template_auto_drafts_theme_versions', $last_auto_drafts_versions );
+}
+add_action( 'wp_loaded', 'gutenberg_create_template_auto_drafts' );
```
This only runs for the requested template - not for every template on every request. Ie. if you are requesting the `single` template then the auto-drafts will be created. But if you never even open the `single` template then you won't have it in the database.

Addison-Stavlo replied:
Ohh shoot I see...  It adds the filter for every type but that filters only run individually when a specific template is requested?

carlomanf replied:
Actually, the part of `gutenberg_find_template_post_and_parts` that creates the auto draft only runs inside an if block:
```
if ( is_admin() || defined( 'REST_REQUEST' ) )
```

so no auto-drafts are ever created on front-end requests.

carlomanf replied:
Actually, the part of `gutenberg_find_template_post_and_parts` that creates the auto draft only runs inside an if block:
```
if ( is_admin() || defined( 'REST_REQUEST' ) )
```

so no auto-drafts are ever created on front-end requests.

carlomanf replied:
Actually, the part of `gutenberg_find_template_post_and_parts` that creates the auto draft only runs inside an if block:
```
if ( is_admin() || defined( 'REST_REQUEST' ) )
```

so no auto-drafts are ever created on front-end requests.

Addison-Stavlo commented:
May also be interested in #26650 - @youknowriad is also looking into the template part auto drafting stuff currently.

youknowriad commented:
It seems we should hold on this one for a bit until #26650 lands and simplifies all this logic.

carlomanf commented:
> It seems we should hold on this one for a bit until #26650 lands and simplifies all this logic.

@youknowriad that one appears to be about template parts, but this one is about templates, or do you anticipate extending that one to cover templates as well?

youknowriad commented:
> that one appears to be about template parts, but this one is about templates, or do you anticipate extending that one to cover templates as well?

Oh right, I missed that part. I'm thinking of taking a look at templates though but I don't know if there will be a lot of changes there. So actually, feel free to proceed here and I'll just check whatever state we'll have.

youknowriad commented:
Actually, I have now updated the PR to also introduce the same "sync" function for templates.

aristath commented:
Just curious.... Why not just store the `filemtime()` as post-meta in the template/template-part in the db, and then reconciliate if the current file's `filemtime` is newer than the stored one?

noahtallen commented:
> Why not just store the filemtime() as post-meta in the template/template-part in the db, and then reconciliate if the current file's filemtime is newer than the stored one

Interesting! I wonder if this would work well with distributed databases/servers 🤔 But it does seem useful for development

carlomanf commented:
> Just curious.... Why not just store the `filemtime()` as post-meta in the template/template-part in the db, and then reconciliate if the current file's `filemtime` is newer than the stored one?

Is there any particular advantage to that over the current approach? It would be easily break if the db were manually modified and it would also complicate changing themes. The current approach by @david-szabo97 seems to work quite well.

aristath commented:
After thinking about this some more, using post-meta would probably be bad for performance... The suggested implementation with the separate option containing versions of all templates feels a bit weird (why save a post's attribute - like its version - in an option instead of the post itself), but at the same time it will perform better.
However, using `filemtime` instead of the theme's version as the template-version would make sense as it can accommodate not only theme-updates, but also manual template edits etc with ease :thinking:

carlomanf commented:
> The suggested implementation with the separate option containing versions of all templates feels a bit weird (why save a post's attribute - like its version - in an option instead of the post itself)

It's not doing that. It's saving the version of the theme, not the version of the post. All theme-related data is stored in the options table. Doesn't seem weird to me.

> However, using `filemtime` instead of the theme's version as the template-version would make sense as it can accommodate not only theme-updates, but also manual template edits etc with ease 🤔

True, but would anyone ever need to edit an FSE template directly?

david-szabo97 commented:
> True, but would anyone ever need to edit an FSE template directly?

Only a developer or theme designer would do that. I don't think a normal user would ever edit a template file directly. Right now this PR syncs the templates when the site editor is opened. After all, I think it's unnecessary and only relevant to developers, so why waste the resources for normal users. How about removing that part or gating it behind a flag that can be toggled in the Gutenberg -> Experiments menu?

aristath commented:
fair point

noahtallen commented:
> How about removing that part or gating it behind a flag that can be toggled in the Gutenberg -> Experiments menu?

Maybe adding a wp-config option for PHP? Perhaps `define( 'BLOCK_THEME_DEV_MODE', true)`? (and always poll from the filesystem directly on that case?) Just an idea.

I do think it's important to support "live updates" of template files for devs, since otherwise you'd have to bump the theme version every time you want to tweak the template file :p

youknowriad commented:
now that the sync PR is merged, should we move the logic here to the "sync" function (there's actually a comment where this might go)

youknowriad commented:
So If I'm reading this right, the sync is done on each page load.  (`wp_loaded`). Do you think we can just rely on the existing synchronization flows and just add the check for content change or date change here https://github.com/WordPress/gutenberg/blob/master/lib/templates-sync.php#L48 instead of adding new hooks?

david-szabo97 commented:
The function will run on every page load. But the sync logic only happens when the current theme or parent theme's version changed. If `BLOCK_THEME_DEV_MODE` constant is true then it syncs them on every load. This is only for developers.

youknowriad commented:
But why the current sync function is not enough to add these checks?

david-szabo97 commented:
https://github.com/WordPress/gutenberg/blob/master/lib/templates-sync.php#L48 This is a separate logic that's unrelated to this PR. I think we had this implemented before: https://github.com/WordPress/gutenberg/pull/26650/files#diff-9f399722fe93f7131d23a3944e030b88420d66461c7b776e7b153bafd3ae7100L165-L193

> But why the current sync function is not enough to add these checks?

I'm fine with running this sync logic when requesting the templates. But I find syncing on `wp_loaded` a better approach. A might not open the site editor after updating the current theme. If we only sync on templates REST request then the templates stay out of sync till the site editor is opened.

youknowriad commented:
> If we only sync on templates REST request then the templates stay out of sync till the site editor is opened.

They stay out of sync until the moment they're needed right? Maybe that's not true but in that case it means we have a bug in master too.

The thing that I don't understand here is that syncing means both: creating auto-drafts for new templates and updating these same auto-drafts for existing ones, I don't understand why these two things need to be in separate places.

david-szabo97 commented:
> They stay out of sync until the moment they're needed right? Maybe that's not true but in that case it means we have a bug in master too.

Yes, we did have this bug in master as well. We didn't have a clear picture of how it should work back then.

From a user perspective, I think it's friendlier to see the template updates right after we update the theme, rather than after opening the site editor. 

**Example: User uses the default template for `front-page`**
- **Case A:** If we sync on `templates` REST request then the user will only see the new `front-page` (both on frontend and backend) after the site-editor is opened
- **Case B:** If we sync on `wp_loaded` by checking whether we have a new theme version or not, the user would see the new `front-page` right after the theme update

IMO, Case B is better in this case, but I might overthink it 😄 

> The thing that I don't understand here is that syncing means both: creating auto-drafts for new templates and updating these same auto-drafts for existing ones, I don't understand why these two things need to be in separate places.

One of us is confused by the purpose of this PR 😄 And now I can't figure out it's you or me. 😅  Basically, we have two separate logic. The "create and update template auto-drafts" logic was cleaned up by you in your [PR](https://github.com/WordPress/gutenberg/pull/26650). The logic is there, but we need to trigger it somehow, that's where this PR comes in. We trigger the sync logic - from your PR - based on specific conditions (when the current or parent theme is updated) in this PR.

youknowriad commented:
For me the current PR was not about triggering the  "create and update template auto-drafts" but more about "update the modified template auto-drafts" and the triggering was already there.

It's fine if we want to change when the syncing happens (the trigger), but the "update template auto-drafts" should happen  in the same place where we create the missing auto-drafts. Here IMO https://github.com/WordPress/gutenberg/blob/master/lib/templates-sync.php#L48

youknowriad commented on docs/architecture/fse-templates.md:
```diff
@@ -30,6 +30,13 @@ The synchronization is important for two different flows:
  - When rendering a template (sometimes referred to as "resolving a template"): this is the algorithm that WordPress follows to traverse the template hierarchy and find the right template to render for the current page being loaded.
  - When exporting a block-based theme, we need to export all its templates back as files. The synchronizaion is required in order to simplify the operation and only export the CPT templates.
 
+Synchronization is triggered in the following situations:
```
should we remove the above list then?

david-szabo97 replied:
Made a couple of changes so reverted my doc changes.

youknowriad commented on lib/templates.php:
```diff
@@ -259,23 +259,62 @@ function filter_rest_wp_template_query( $args, $request ) {
+	}
+	update_option( 'gutenberg_last_template_auto_drafts_theme_versions', $last_auto_drafts_versions );
+}
+add_action( 'wp_loaded', 'gutenberg_synchronize_theme_templates_on_version_change' );
```
Should we move this code to `tempate-sync.php` file?

youknowriad commented on lib/templates.php:
```diff
@@ -259,23 +259,62 @@ function filter_rest_wp_template_query( $args, $request ) {
-	if ( null !== $dispatch_result ) {
-		return $dispatch_result;
+function gutenberg_synchronize_theme_templates_on_version_change() {
+	if ( defined( 'BLOCK_THEME_DEV_MODE' ) && BLOCK_THEME_DEV_MODE ) {
```
I'm not sure a constant is the best way to go about it, there's a ton of use-cases where the user might edit the files without having this constant:

 - a random plugin might do it
 - the user could do it in the "theme editor"

I personally feel like relying on the file's modified date like suggested by @aristath is probably better. 
I'm also concerned about the number of issues theme authors will raise where they won't be aware of this constant.

david-szabo97 replied:
Good point

youknowriad reviewed (approved):
The code is looking good for me. Maybe we want to change the "when" sync happens in the architecture docs?

I didn't test it though.

david-szabo97 commented:
Reverted lots of stuff, changed title and description. Basically, this PR now adds a block of code that checks if the files were modified since the last check or not.

youknowriad commented:
@david-szabo97 I do wonder though whether your original idea of running the synchronization on wp_loaded is better than what we have now to avoid hidden bugs or flows where we'd forget to run it.

I guess performance could be a potential problem but since we're doing the date checks, maybe it's not?

david-szabo97 commented:
Yeah, I was thinking about keeping the sync on `wp_loaded`. I also had concerns about date checks causing perf problems. But looks like it isn't that bad. It's metadata provided by the OS so we don't need to read the file.  Looking at `filemtime` benchmarks it shouldn' be a problem.

Copons commented:
I'm concerned about the performance impact of running the sync at every load, but I've also just spent an absurd amount of time tracking down when **one** of the syncs happens.

I wonder if by being performance-conscious we are also making things extremely complicated to debug.

I'm no performance expert, but I'm inclined to think that with the time checks we are in the clear.

aristath commented on lib/templates-sync.php:
```diff
@@ -87,12 +90,31 @@ function _gutenberg_synchronize_theme_templates( $template_type ) {
+		$themes[] = get_template();
 	}
 
+	// Get file paths for all theme supplied template that changed since last check.
+	$template_files = array();
+	$option_name    = 'gutenberg_last_synchronize_theme_' . $template_type . '_checks';
+	$last_checks    = get_option( $option_name, array() );
+	$current_time   = time();
+	foreach ( $themes as $theme_slug ) {
+		$theme      = wp_get_theme( $theme_slug );
+		$last_check = isset( $last_checks[ $theme_slug ] ) ? $last_checks[ $theme_slug ] : 0;
+
+		$theme_template_files = _gutenberg_get_template_paths( $theme->get_stylesheet_directory() . '/' . $template_base_paths[ $template_type ] );
+		foreach ( $theme_template_files as $template_file ) {
+			if ( filemtime( $template_file ) > $last_check ) {
+				$template_files[] = $template_file;
+			}
+		}
+
+		$last_checks[ $theme_slug ] = $current_time;
+	}
+	update_option( $option_name, $last_checks );
+
```
This code updates the option depending on the file dates, but what does that option actually do? I don't see it used anywhere... 🤔

david-szabo97 replied:
Used on Line 105 then Line 109.
```php
$last_check = isset( $last_checks[ $theme_slug ] ) ? $last_checks[ $theme_slug ] : 0;
.....
if ( filemtime( $template_file ) > $last_check ) {
```

The option stores the last date of synchronization check. It is an associative array where K is the theme stylesheet name and V is the time of the last check.

aristath replied:
Ah OK, I see it now, based on that option the array of `$template_files` is build and then that array goes to the other `foreach` loop. Gotcha 👍

david-szabo97 commented on lib/templates-sync.php:
```diff
@@ -87,12 +90,31 @@ function _gutenberg_synchronize_theme_templates( $template_type ) {
-	// Get file paths for all theme supplied template.
-	$template_files = _gutenberg_get_template_paths( get_stylesheet_directory() . '/' . $template_base_paths[ $template_type ] );
+	$themes = array( get_stylesheet() );
 	if ( is_child_theme() ) {
-		$template_files = array_merge( $template_files, _gutenberg_get_template_paths( get_template_directory() . '/' . $template_base_paths[ $template_type ] ) );
+		$themes[] = get_template();
```
Smart! Thanks!

david-szabo97 commented:
Now synchronization is only called at two places: when `wp_loaded` and before site export.

aristath commented:
Just curious (and mostly thinking out loud)... Now that we actually use `filemtime`, could we not just compare that against the auto-draft's update-datetime instead of creating a new option? 🤔

Copons commented:
> Just curious (and mostly thinking out loud)... Now that we actually use `filemtime`, could we not just compare that against the auto-draft's update-datetime instead of creating a new option? 🤔

I guess we could check the most recent auto-draft's update time... 🤔

david-szabo97 commented:
The new option is a single query, whereas checking the auto-drafts would require a single query for each template, plus we don't need the whole WP_Post object in this case.

aristath commented:
> The new option is a single query, whereas checking the auto-drafts would require a single query for each template, plus we don't need the whole WP_Post object in this case.

Fair point. The option will indeed be more performant.

noahtallen commented on lib/templates-sync.php:
```diff
@@ -86,12 +89,27 @@ function _gutenberg_synchronize_theme_templates( $template_type ) {
 		'template'      => 'block-templates',
 		'template-part' => 'block-template-parts',
 	);
+	$themes              = array_unique( array( get_stylesheet(), get_template() ) );
```
What about using this:

```php
$themes = array_unique( array( get_stylesheet_directory(), get_template_directory() ) );
```

Then we could avoid the call to `wp_get_theme` and `get_stylesheet_directory`. Then we could just use the directory as the slug identifier, or just the basename of the patth.

Probably not a big deal either way, given that we're doing a lot of filesystem stuff here anyways.

david-szabo97 replied:
`basename` wouldn't return correct `stylesheet` in case of "nested" themes. ( for example themes/seedlet-blocks would become seedlet-blocks)

I used an assoc array though to avoid creating WP_Theme objects.

noahtallen commented on lib/templates-sync.php:
```diff
@@ -106,3 +124,12 @@ function _gutenberg_synchronize_theme_templates( $template_type ) {
+	_gutenberg_synchronize_theme_templates( 'template' );
+	_gutenberg_synchronize_theme_templates( 'template_part' );
+}
+add_action( 'wp_loaded', 'gutenberg_synchronize_theme_templates_on_load' );
```
Since the `templates-sync.php` file is loaded all the time (regardless of whether an FSE theme is active), we should probably make sure that `gutenberg_synchronize_theme_templates_on_load` only runs if FSE is enabled. (e.g. shortcut if `gutenberg_is_fse_theme` is false)

david-szabo97 replied:
Gotcha! Thanks!

noahtallen reviewed (commented):
This overall looks awesome, thanks for working on this!

Unfortunately, I am noticing that some template parts are not loading in the editor. They _are_ loading on the front end of the site with the correct template, but they don't load in the site editor after a few loads.

<img width="1493" alt="Screen Shot 2020-11-12 at 4 18 41 PM" src="https://user-images.githubusercontent.com/6265975/99012530-7c91ed80-2503-11eb-8043-a6a984f6a3e9.png">

Here are the details:
1. twenty twentyone blocks _does_ work. This is loaded as normal via wp-env.
2. I have the [Automattic/themes](https://github.com/Automattic/themes/) mapped to `wp-content/themes/themes`.
3. Neither seedlet blocks or ibis (both from the above repo) will load template parts in the site editor.

I investigated a bit, and it is resolving the correct directory for the template parts:

```php
$theme_slug = get_stylesheet();
// $theme_slug = "themes/seedlet-blocks"

$theme = wp_get_theme( $theme_slug );

$theme_dir = $theme->get_stylesheet_directory();
// $theme_dir = "/var/www/html/wp-content/themes/themes/seedlet-blocks"
```

And that directory exists and contains the "block-template-parts" directory. So it seems like the code _should_ be working. I really can't tell why twenty twenty one blocks is working but not the other two.

david-szabo97 commented:
> Unfortunately, I am noticing that some template parts are not loading in the editor. They are loading on the front end of the site with the correct template, but they don't load in the site editor after a few loads.

Thanks for testing with other themes, I totally forgot to do that 😄  This [commit](https://github.com/WordPress/gutenberg/pull/26383/commits/fffcee0967833fca4b136f409ceb4be63fc1a18c) should fix the bug. Can you give it a check?

noahtallen commented on lib/templates-sync.php:
```diff
@@ -86,23 +89,60 @@ function _gutenberg_synchronize_theme_templates( $template_type ) {
 		);
-		_gutenberg_create_auto_draft_for_template( $template_post_types[ $template_type ], $slug, wp_get_theme()->get_stylesheet(), $content );
+		_gutenberg_create_auto_draft_for_template( $template_post_types[ $template_type ], $slug, $theme, $content );
 	}
```
Let's move `update_option( $option_name, $last_checks );` to the very end of this function. That way, if something like `_gutenberg_create_auto_draft_for_template` errors out, hopefully the option would not be updated. I noticed that if that function has an error, then the templates are not created, but the option is updated, so it doesn't try again the next time.

Maybe we should also make sure that `_gutenberg_create_auto_draft_for_template` doesn't return an error

david-szabo97 replied:
Great point about updating the option. Moved it to the end of the function.

I don't think we can make sure `_gutenberg_create_auto_draft_for_template` doesn't return any error. It's doing database queries so I'm not sure what we could do in this case.

noahtallen commented on lib/templates-sync.php:
```diff
@@ -86,23 +89,60 @@ function _gutenberg_synchronize_theme_templates( $template_type ) {
+	}
+
+	_gutenberg_synchronize_theme_templates( 'template' );
+	_gutenberg_synchronize_theme_templates( 'template_part' );
```
```suggestion
	_gutenberg_synchronize_theme_templates( 'template-part' );
```

the associative arrays have `template-part` as keys

david-szabo97 replied:
What a silly mistake.... 😠  Sorry about that

noahtallen reviewed (changes requested):
It's weird, for some reason it's still not working. Specifically, the `index` template won't load its template parts. The `singular` template does, and I can select the template parts from the sidebar. _Edit: actually, the singular template won't either after cleaning the database._

<img width="1832" alt="Screen Shot 2020-11-13 at 2 57 57 PM" src="https://user-images.githubusercontent.com/6265975/99128862-d7d5e580-25c0-11eb-9aba-c86aef99358a.png">

youknowriad commented:
how ready is this PR? It would be good to include in the RC release today.

david-szabo97 commented:
Fixed the errors reported by Noah above. Did a few tests and seems to work fine. If you could do some tests and report back that would be great. I'm trying to get E2E tests to pass.

youknowriad commented:
@david-szabo97 There's a lot of e2e  test failures on master now which I'm working on. Most of the failures in the PR should be unrelated but I'll you know once I make progress so you can rebase.

jorgefilipecosta committed:
Site Editor: Create auto-drafts for modified files only

Copons commented:
I've stumbled upon a potential issue, that might or might not be a big deal, depending on how "in depth" a site is used.

Recently I've been working a lot with templates and template parts, and to keep things safe while switching branches, I got used to delete all `wp_template` and `wp_template_part` posts all the times.

As it turned out, I've tested this branch, then changed to work on something else, then deleted all templates and parts, and switched back on this branch.
What happened was that no templates and parts were created.

This is because the time check options were created during my original test.
When I came back to this branch _without any templates and parts posts_, **`$last_check` was set to a timestamp more recent than the theme files modification date** (as I normally don't wipe the theme).

For new themes activations, this PR relies on the _absence_ of time check options for the given theme, defaulting to 0, which is safely older than anything:
```php
$last_check = isset( $last_checks[ $theme_slug ] ) ? $last_checks[ $theme_slug ] : 0;
```

This was also confirmed by the fact that once I deleted the time check options too, the template and parts were created successfully.

In other words: **if a user deletes all `wp_template` (and `_part`) posts, they also need to delete the `gutenberg_last_synchronize_theme_{ type }_checks` options, otherwise the site won't create the auto-drafts anymore, leaving the site broken.**

---

I fully recognize that this is not a normal flow, but I also can't help but wonder if the time check options might unintentionally hinder what is supposed to be a relatively safe operation, without possibility of recover (unless one is aware of the options).

This will be aggravated if we end up displaying auto-draft in the templates and parts wp-admin lists (as proposed in #26636). (As a side note: the intention with showing auto-drafts there is to help with the development, rather than being a production feature)

Am I overestimating this potential issue?

aristath commented:
I think it's an edge case... but perhaps we could delete the stored timestamps if the file doesn't exist? :thinking:

Copons commented:
> I think it's an edge case... but perhaps we could delete the stored timestamps if the file doesn't exist? 🤔

@aristath Unfortunately it's the other way around 😅 

The files are there, untouched.
What I deleted are the templates and parts **posts**!

Basically we would need to:
- Is the file more recent than the last check?
  - Yes: update the template post. 👍 
  - No. Is there a template post in the DB?
    - Yes: good to go! 👍 
    - No: create a template post for that file.

Which... I'm pretty sure it would remove all benefits of this PR. 😄

david-szabo97 committed:
Clear last checks option after trashing or deleting templates and template parts

david-szabo97 committed:
Fix error

Copons commented:
With the 6e8bb9d change, the issue seems fixed.

Just for additional context, I have found a definitely **not** edge case that would have triggered this issue (which, again, is now fixed 🙂):

- Let's assume we have a clean site with twenty-twentyone-blocks.
- Tun the templates sync by opening the front end or the Site Editor.
- Open the Site Editor: it should be on the Front Page template.
- Customize the template and save.
- Open the Templates list (`/wp-admin/edit.php?post_type=wp_template`): there should be a Front Page template.
- Trash it.
- Open the Site Editor again: now it's on the Index template (which is the fallback if Front Page is missing).
- Look in the sidebar: no Front Page template anymore.

This was caused by the same issue: we only compared the sync date with the file date, but didn't consider that a template might be deleted.

6e8bb9d should fix most use cases. I guess power users should be also able to delete a site option, if needs arise. 🤔

vindl commented on lib/templates-sync.php:
```diff
@@ -86,23 +89,86 @@ function _gutenberg_synchronize_theme_templates( $template_type ) {
+ * @param WP_Post $post WP_Post instance of the deleted post.
+ */
+function gutenberg_clear_synchronize_last_checks_after_delete( $postid, $post ) {
+	if ( 'wp_template' !== $post->post_type || 'wp_template_part' !== $post->post_type ) {
```
Should this be `'wp_template' === $post->post_type`?

youknowriad replied:
do we have a PR to fix the issues introduced here so I can backport them to 9.4?

vindl replied:
I started one here https://github.com/WordPress/gutenberg/pull/27068.
